tt_metal/common/core_coord.hpp

```diff
@@ -0,0 +1,253 @@
+
+    std::string str() const;
+};
+
+constexpr inline bool operator==(const RelativeCoreCoord &a, const RelativeCoreCoord &b) {
+    return a.x == b.x && a.y == b.y;
+}
+
+constexpr inline bool operator!=(const RelativeCoreCoord &a, const RelativeCoreCoord &b) { return !(a == b); }
+
+CoreCoord get_core_coord_from_relative(const RelativeCoreCoord &in, const CoreCoord &grid_size);
+
+struct CoreRange {
+    CoreCoord start_coord;
+    CoreCoord end_coord;
+    CoreRange(const CoreCoord &point);
+
+    CoreRange(const CoreCoord &start_coord, const CoreCoord &end_coord);
+
+    CoreRange(const CoreRange &other) = default;
+    CoreRange &operator=(const CoreRange &other) = default;
+    CoreRange(CoreRange &&other) = default;
+    CoreRange &operator=(CoreRange &&other) = default;
+
+    std::optional<CoreRange> intersects(const CoreRange &other) const;
+
+    bool adjacent(const CoreRange &other) const;
+
+    bool contains(const CoreRange &other) const;
+
+    bool contains(const CoreCoord &other) const;
+
+    // Merge lined-up (in x or y dimension) intersecting/adjacent rectangles
+    std::optional<CoreRange> merge(const CoreRange &cr) const;
+
+    std::string str() const;
+
+    size_t size() const;
+
+    CoreCoord grid_size() const;
+
+    class CoreIterator {
+       public:
+        CoreIterator(const CoreCoord &current, const CoreRange &core_range);
+
+        CoreCoord &operator*();
+
+        CoreIterator &operator++();
+
+        bool operator==(const CoreIterator &other) const;
+
+        bool operator!=(const CoreIterator &other) const;
+
+       private:
+        CoreCoord current_;
+        const CoreRange &range_;
+    };
+
+    CoreIterator begin() const;
+
+    CoreIterator end() const;
+};
+
+constexpr bool operator==(const CoreRange &a, const CoreRange &b) {
+    return a.start_coord == b.start_coord && a.end_coord == b.end_coord;
+}
+
+constexpr bool operator!=(const CoreRange &a, const CoreRange &b) { return !(a == b); }
+
+constexpr bool operator<(const CoreRange &left, const CoreRange &right) {
+    return (
+        left.start_coord < right.start_coord ||
+        (left.start_coord == right.start_coord && left.end_coord < right.end_coord));
+}
+
+template <>
+struct fmt::formatter<CoreRange> {
+    constexpr auto parse(format_parse_context &ctx) -> format_parse_context::iterator { return ctx.end(); }
+
+    auto format(const CoreRange &core_range, format_context &ctx) const -> format_context::iterator;
+};
+
+class CoreRangeSet {
+   public:
+    CoreRangeSet(const std::vector<CoreRange> &core_ranges);
+
+    CoreRangeSet(const std::set<CoreRange> &core_ranges);
+
+    CoreRangeSet(const CoreRange &core_range);
+
+    CoreRangeSet() = default;
+
+    friend void swap(CoreRangeSet &first, CoreRangeSet &second);
+
+    CoreRangeSet(const CoreRangeSet &other);
+
+    CoreRangeSet &operator=(const CoreRangeSet &other);
+
+    CoreRangeSet(CoreRangeSet &&other);
+
+    CoreRangeSet &operator=(CoreRangeSet &&other);
```

Comment:
Can you add a move constructor from vector?
```cpp
CoreRangeSet(std::vector<CoreRange> &&core_ranges) noexcept;

...

CoreRangeSet::CoreRangeSet(std::vector<CoreRange> &&core_ranges) noexcept : ranges_(std::move(core_ranges)) {}
```

Reply:
Done